File: packages/typescript-sdk/src/resources/agency.ts
import { HTTPClient } from '../client/http';

// ── Types ──────────────────────────────────────────────────────────────

export interface Agency {
  agency_id: string;
  name: string;
  plan_tier?: string;
  wallet_credits: number;
  per_brand_caps_enabled: boolean;
  max_brands?: number;
  created_at: string;
  updated_at: string;
}

export interface Brand {
  brand_id: string;
  agency_id?: string;
  name: string;
  industry?: string;
  logo_url?: string;
  monthly_credit_cap?: number;
  archived: boolean;
  created_at: string;
  updated_at: string;
}

export interface AgencyMember {
  agency_member_id: string;
  agency_id: string;
  user_id?: string;
  email?: string;
  role: 'admin' | 'member';
  status: 'active' | 'pending' | 'conflicted';
  assigned_brand_ids?: string[];
  created_at: string;
}

export interface RosterCard extends Brand {
  pending_approvals: number;
  scheduled_today: number;
  credits_consumed_this_month: number;
  last_activity?: string;
}

// ── Request Types ──────────────────────────────────────────────────────

export interface CreateAgencyRequest {
  name: string;
  plan_tier?: string;
}

export interface UpdateAgencyRequest {
  name: string;
}

export interface CreateBrandRequest {
  name: string;
  industry?: string;
  logo_url?: string;
  monthly_credit_cap?: number;
}

export interface DuplicateBrandRequest {
  template_brand_id: string;
  name: string;
  industry?: string;
}

export interface UpdateBrandRequest {
  name?: string;
  industry?: string;
  logo_url?: string;
  monthly_credit_cap?: number;
}

export interface InviteMemberRequest {
  email: string;
  role?: 'admin' | 'member';
}

export interface TopUpWalletRequest {
  credits: number;
}

// ── Resource ───────────────────────────────────────────────────────────

export class AgencyResource {
  constructor(private http: HTTPClient) {}

  // ── Agency ──────────────────────────────────────────────────────────

  /**
   * Create a new agency (caller becomes admin)
   */
  async create(request: CreateAgencyRequest): Promise<{ agency: Agency }> {
    return this.http.post<{ agency: Agency }>('/agency', request);
  }

  /**
   * Get caller's agency
   */
  async get(): Promise<{ agency: Agency | null }> {
    return this.http.get<{ agency: Agency | null }>('/agency');
  }

  /**
   * Update agency name (admin only)
   */
  async update(request: UpdateAgencyRequest): Promise<{ agency: Agency }> {
    return this.http.patch<{ agency: Agency }>('/agency', request);
  }

  /**
   * Upgrade solo SME to agency, bringing existing brand in
   */
  async upgradeSoloToAgency(request: CreateAgencyRequest): Promise<{ agency: Agency & { migrated_brand_id: string } }> {
    return this.http.post<{ agency: Agency & { migrated_brand_id: string } }>('/agency/upgrade', request);
  }

  /**
   * Toggle per-brand credit caps (admin only)
   */
  async updateSettings(perBrandCapsEnabled: boolean): Promise<{ agency: { per_brand_caps_enabled: boolean } }> {
    return this.http.patch<{ agency: { per_brand_caps_enabled: boolean } }>(
      '/agency/settings',
      null,
      { params: { per_brand_caps_enabled: perBrandCapsEnabled } }
    );
  }

  // ── Brands ──────────────────────────────────────────────────────────

  /**
   * List all brands accessible to caller (brand-switcher list)
   */
  async listBrands(): Promise<{ brand: Brand[] }> {
    return this.http.get<{ brand: Brand[] }>('/agency/brands');
  }

  /**
   * Add a new brand to agency (admin only)
   */
  async createBrand(request: CreateBrandRequest): Promise<{ brand: Brand }> {
    return this.http.post<{ brand: Brand }>('/agency/brands', request);
  }

  /**
   * Duplicate brand from existing template (admin only)
   */
  async duplicateBrand(request: DuplicateBrandRequest): Promise<{ brand: Brand }> {
    return this.http.post<{ brand: Brand }>('/agency/brands/duplicate', request);
  }

  /**
   * Update brand details (admin only)
   */
  async updateBrand(brandId: string, updates: UpdateBrandRequest): Promise<{ brand: Brand }> {
    return this.http.patch<{ brand: Brand }>(`/agency/brands/${brandId}`, updates);
  }

  /**
   * Archive a brand (admin only)
   */
  async archiveBrand(brandId: string): Promise<{ brand: { brand_id: string; archived: boolean } }> {
    return this.http.delete<{ brand: { brand_id: string; archived: boolean } }>(`/agency/brands/${brandId}`);
  }

  /**
   * Get agency roster - all accessible brands with at-a-glance status
   */
  async getRoster(): Promise<{ roster_brand: RosterCard[] }> {
    return this.http.get<{ roster_brand: RosterCard[] }>('/agency/roster');
  }

  // ── Members ─────────────────────────────────────────────────────────

  /**
   * List agency members (admin only)
   */
  async listMembers(): Promise<{ agency_member: AgencyMember[] }> {
    return this.http.get<{ agency_member: AgencyMember[] }>('/agency/members');
  }

  /**
   * Invite member to agency (admin only)
   */
  async inviteMember(request: InviteMemberRequest): Promise<{ agency_member: AgencyMember & { message: string } }> {
    return this.http.post<{ agency_member: AgencyMember & { message: string } }>('/agency/members', request);
  }

  /**
   * Remove member from agency (admin only)
   */
  async removeMember(memberId: string): Promise<{ agency_member: { agency_member_id: string; removed: boolean } }> {
    return this.http.delete<{ agency_member: { agency_member_id: string; removed: boolean } }>(
      `/agency/members/${memberId}`
    );
  }

  /**
   * Assign brand access to member (admin only)
   */
  async assignBrandToMember(
    memberId: string,
    brandId: string
  ): Promise<{ member_brand_access: { agency_member_id: string; brand_id: string; assigned: boolean } }> {
    return this.http.post<{ member_brand_access: { agency_member_id: string; brand_id: string; assigned: boolean } }>(
      `/agency/members/${memberId}/brands/${brandId}`
    );
  }

  /**
   * Remove brand access from member (admin only)
   */
  async unassignBrandFromMember(
    memberId: string,
    brandId: string
  ): Promise<{ member_brand_access: { agency_member_id: string; brand_id: string; unassigned: boolean } }> {
    return this.http.delete<{ member_brand_access: { agency_member_id: string; brand_id: string; unassigned: boolean } }>(
      `/agency/members/${memberId}/brands/${brandId}`
    );
  }

  // ── Wallet ──────────────────────────────────────────────────────────

  /**
   * Top up agency wallet (admin only)
   */
  async topupWallet(request: TopUpWalletRequest): Promise<{ agency_wallet: { wallet_credits: number } }> {
    return this.http.post<{ agency_wallet: { wallet_credits: number } }>('/agency/wallet/topup', request);
  }

  // ── Reports ─────────────────────────────────────────────────────────

  /**
   * Get portfolio dashboard - agency-wide metrics (admin only)
   */
  async getPortfolioReport(): Promise<{
    portfolio_report: {
      agency_id: string;
      wallet_credits: number;
      total_credits_consumed_this_month: number;
      total_posts_published: number;
      brand_count: number;
      per_brand: Array<{
        brand_id: string;
        name: string;
        credits_consumed_this_month: number;
        posts_published: number;
        pending_approvals: number;
        needs_attention: boolean;
      }>;
    };
  }> {
    return this.http.get('/agency/reports/portfolio');
  }

  /**
   * Get per-brand client report
   */
  async getBrandReport(brandId: string): Promise<{
    brand_report: {
      brand_id: string;
      name: string;
      posts_published: number;
      credits_consumed_this_month: number;
      top_posts: Array<{ generated_title: string; primary_keyword: string }>;
    };
  }> {
    return this.http.get(`/agency/reports/brand/${brandId}`);
  }
}

File: packages/python-sdk/src/urisocial/resources/agency.py
"""Agency management resource for URI Social SDK"""

from typing import Dict, Any, List, Optional
from ..http_client import HTTPClient


class AgencyResource:
    """Agency account management - multi-brand agencies"""

    def __init__(self, http: HTTPClient):
        self._http = http

    # ── Agency ──────────────────────────────────────────────────────────

    def create(self, name: str, plan_tier: Optional[str] = None) -> Dict[str, Any]:
        """
        Create a new agency (caller becomes admin)

        Args:
            name: Agency name
            plan_tier: Optional plan tier (e.g., 'starter', 'growth', 'enterprise')

        Returns:
            Created agency object
        """
        payload = {"name": name}
        if plan_tier:
            payload["plan_tier"] = plan_tier
        return self._http.post("/agency", json=payload)

    def get(self) -> Dict[str, Any]:
        """
        Get caller's agency

        Returns:
            Agency object or None if user doesn't belong to an agency
        """
        return self._http.get("/agency")

    def update(self, name: str) -> Dict[str, Any]:
        """
        Update agency name (admin only)

        Args:
            name: New agency name

        Returns:
            Updated agency object
        """
        return self._http.patch("/agency", json={"name": name})

    def upgrade_solo_to_agency(self, name: str, plan_tier: Optional[str] = None) -> Dict[str, Any]:
        """
        Upgrade solo SME to agency, bringing existing brand in

        Args:
            name: Agency name
            plan_tier: Optional plan tier

        Returns:
            Created agency object with migrated_brand_id
        """
        payload = {"name": name}
        if plan_tier:
            payload["plan_tier"] = plan_tier
        return self._http.post("/agency/upgrade", json=payload)

    def update_settings(self, per_brand_caps_enabled: bool) -> Dict[str, Any]:
        """
        Toggle per-brand credit caps (admin only)

        Args:
            per_brand_caps_enabled: Enable/disable per-brand caps

        Returns:
            Updated settings
        """
        return self._http.patch("/agency/settings", params={"per_brand_caps_enabled": per_brand_caps_enabled})

    # ── Brands ──────────────────────────────────────────────────────────

    def list_brands(self) -> Dict[str, Any]:
        """
        List all brands accessible to caller (brand-switcher list)

        Returns:
            List of brand objects
        """
        return self._http.get("/agency/brands")

    def create_brand(
        self,
        name: str,
        industry: Optional[str] = None,
        logo_url: Optional[str] = None,
        monthly_credit_cap: Optional[float] = None,
    ) -> Dict[str, Any]:
        """
        Add a new brand to agency (admin only)

        Args:
            name: Brand name
            industry: Optional industry
            logo_url: Optional logo URL
            monthly_credit_cap: Optional monthly credit cap

        Returns:
            Created brand object
        """
        payload = {"name": name}
        if industry:
            payload["industry"] = industry
        if logo_url:
            payload["logo_url"] = logo_url
        if monthly_credit_cap is not None:
            payload["monthly_credit_cap"] = monthly_credit_cap
        return self._http.post("/agency/brands", json=payload)

    def duplicate_brand(
        self,
        template_brand_id: str,
        name: str,
        industry: Optional[str] = None,
    ) -> Dict[str, Any]:
        """
        Duplicate brand from existing template (admin only)

        Args:
            template_brand_id: ID of brand to duplicate from
            name: New brand name
            industry: Optional industry

        Returns:
            Created brand object
        """
        payload = {"template_brand_id": template_brand_id, "name": name}
        if industry:
            payload["industry"] = industry
        return self._http.post("/agency/brands/duplicate", json=payload)

    def update_brand(self, brand_id: str, updates: Dict[str, Any]) -> Dict[str, Any]:
        """
        Update brand details (admin only)

        Args:
            brand_id: Brand ID
            updates: Dictionary of fields to update (name, industry, logo_url, monthly_credit_cap, etc.)

        Returns:
            Updated brand object
        """
        return self._http.patch(f"/agency/brands/{brand_id}", json=updates)

    def archive_brand(self, brand_id: str) -> Dict[str, Any]:
        """
        Archive a brand (admin only)

        Args:
            brand_id: Brand ID to archive

        Returns:
            Archive confirmation
        """
        return self._http.delete(f"/agency/brands/{brand_id}")

    def get_roster(self) -> Dict[str, Any]:
        """
        Get agency roster - all accessible brands with at-a-glance status

        Returns:
            List of roster cards with pending_approvals, scheduled_today, credits_consumed_this_month, etc.
        """
        return self._http.get("/agency/roster")

    # ── Members ─────────────────────────────────────────────────────────

    def list_members(self) -> Dict[str, Any]:
        """
        List agency members (admin only)

        Returns:
            List of agency members with assigned brand IDs
        """
        return self._http.get("/agency/members")

    def invite_member(self, email: str, role: str = "member") -> Dict[str, Any]:
        """
        Invite member to agency (admin only)

        Args:
            email: Email address to invite
            role: Role ('admin' or 'member', default: 'member')

        Returns:
            Created member object with invite status
        """
        return self._http.post("/agency/members", json={"email": email, "role": role})

    def remove_member(self, member_id: str) -> Dict[str, Any]:
        """
        Remove member from agency (admin only)

        Args:
            member_id: Agency member ID

        Returns:
            Removal confirmation
        """
        return self._http.delete(f"/agency/members/{member_id}")

    def assign_brand_to_member(self, member_id: str, brand_id: str) -> Dict[str, Any]:
        """
        Assign brand access to member (admin only)

        Args:
            member_id: Agency member ID
            brand_id: Brand ID to assign

        Returns:
            Assignment confirmation
        """
        return self._http.post(f"/agency/members/{member_id}/brands/{brand_id}")

    def unassign_brand_from_member(self, member_id: str, brand_id: str) -> Dict[str, Any]:
        """
        Remove brand access from member (admin only)

        Args:
            member_id: Agency member ID
            brand_id: Brand ID to unassign

        Returns:
            Unassignment confirmation
        """
        return self._http.delete(f"/agency/members/{member_id}/brands/{brand_id}")

    # ── Wallet ──────────────────────────────────────────────────────────

    def topup_wallet(self, credits: float) -> Dict[str, Any]:
        """
        Top up agency wallet (admin only)

        Args:
            credits: Number of credits to add

        Returns:
            Updated wallet balance
        """
        return self._http.post("/agency/wallet/topup", json={"credits": credits})

    # ── Reports ─────────────────────────────────────────────────────────

    def get_portfolio_report(self) -> Dict[str, Any]:
        """
        Get portfolio dashboard - agency-wide metrics (admin only)

        Returns:
            Portfolio report with wallet_credits, total_credits_consumed_this_month, per_brand breakdown, etc.
        """
        return self._http.get("/agency/reports/portfolio")

    def get_brand_report(self, brand_id: str) -> Dict[str, Any]:
        """
        Get per-brand client report

        Args:
            brand_id: Brand ID

        Returns:
            Brand report with posts_published, credits_consumed_this_month, top_posts, etc.
        """
        return self._http.get(f"/agency/reports/brand/{brand_id}")
